AnkiDroid/src/test/java/com/ichi2/anki/export/ExportDialogFragmentTest.kt

```
@@ -0,0 +1,69 @@
+/*
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation; either version 3 of the License, or (at your option) any later
+ * version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ * PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.ichi2.anki.export
+
+import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.action.ViewActions.click
+import androidx.test.espresso.assertion.ViewAssertions.matches
+import androidx.test.espresso.matcher.RootMatchers.isDialog
+import androidx.test.espresso.matcher.ViewMatchers.Visibility
+import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
+import androidx.test.espresso.matcher.ViewMatchers.withEffectiveVisibility
+import androidx.test.espresso.matcher.ViewMatchers.withId
+import androidx.test.espresso.matcher.ViewMatchers.withText
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.ichi2.anki.R
+import com.ichi2.testutils.AnkiFragmentScenario
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class ExportDialogFragmentTest {
+    @Test
+    fun colLegacyExportCheckbox() {
+        AnkiFragmentScenario.launchInContainer(ExportDialogFragment::class.java).onFragment { fragment ->
+            fragment.view?.let {
```

Why was _fragment.view?._ added, the test should work without this.

---

Oops leftover from when i used a Espresso Test from HelpDialogTest.kt as a reference.
Will remove!

---

```
Subject: [PATCH] Unit Tests for initializeApkgExportUi and initializeNotesExportUi
---
Index: AnkiDroid/src/test/java/com/ichi2/anki/export/ExportDialogFragmentTest.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/test/java/com/ichi2/anki/export/ExportDialogFragmentTest.kt b/AnkiDroid/src/test/java/com/ichi2/anki/export/ExportDialogFragmentTest.kt
--- a/AnkiDroid/src/test/java/com/ichi2/anki/export/ExportDialogFragmentTest.kt	(revision ff8e1e9327f860a28449e6557d840f690c26532f)
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/export/ExportDialogFragmentTest.kt	(date 1735583790601)
@@ -32,38 +32,29 @@
 class ExportDialogFragmentTest {
     @Test
     fun colLegacyExportCheckbox() {
-        AnkiFragmentScenario.launchInContainer(ExportDialogFragment::class.java).onFragment { fragment ->
-            fragment.view?.let {
-                onView(withId(R.id.export_type_selector)).perform(click())
-                onView(withText("Anki Collection Package (.colpkg)")).perform(click())
-                onView(withId(R.id.export_legacy_checkbox_collection))
-                    .inRoot(isDialog())
-                    .check(matches(isDisplayed()))
-            }
-        }
+        AnkiFragmentScenario.launchInContainer(ExportDialogFragment::class.java)
+        onView(withId(R.id.export_type_selector)).perform(click())
+        onView(withText("Anki Collection Package (.colpkg)")).perform(click())
+        onView(withId(R.id.export_legacy_checkbox_collection))
+            .inRoot(isDialog())
+            .check(matches(isDisplayed()))
     }
 
     @Test
     fun apkgLegacyExportCheckbox() {
-        AnkiFragmentScenario.launchInContainer(ExportDialogFragment::class.java).onFragment { fragment ->
-            fragment.view?.let {
-                onView(withId(R.id.export_type_selector)).perform(click())
-                onView(withText("Anki Deck Package (.apkg)")).perform(click())
-                onView(withId(R.id.export_legacy_checkbox_collection))
-                    .inRoot(isDialog())
-                    .check(matches(isDisplayed()))
-            }
-        }
+        AnkiFragmentScenario.launchInContainer(ExportDialogFragment::class.java)
+        onView(withId(R.id.export_type_selector)).perform(click())
+        onView(withText("Anki Deck Package (.apkg)")).perform(click())
+        onView(withId(R.id.export_legacy_checkbox_collection))
+            .inRoot(isDialog())
+            .check(matches(isDisplayed()))
     }
 
     @Test
     fun defaultLegacyExportCheckbox() {
-        AnkiFragmentScenario.launchInContainer(ExportDialogFragment::class.java).onFragment { fragment ->
-            fragment.view?.let {
-                onView(withId(R.id.export_legacy_checkbox_collection))
-                    .inRoot(isDialog())
-                    .check(matches(withEffectiveVisibility(Visibility.GONE)))
-            }
-        }
+        AnkiFragmentScenario.launchInContainer(ExportDialogFragment::class.java)
+        onView(withId(R.id.export_legacy_checkbox_collection))
+            .inRoot(isDialog())
+            .check(matches(withEffectiveVisibility(Visibility.GONE)))
     }
-}
+}
\ No newline at end of file

```

---

With this as a patch the test keeps failing, perhaps the views aren't initialized?

---

Again, the test should pass without the query for the fragment's view. In fact we wouldn't really know if the test checks something, if the fragment's view is null the checks are silently ignored.

You need to figure out why is the test failing. Also, you are in a dialog so all espresso calls should be scoped to the dialog, you're doing this, for some reason, just for the last line?!?!.

---

```
Subject: [PATCH] Unit Tests for initializeApkgExportUi and initializeNotesExportUi
---
Index: AnkiDroid/src/test/java/com/ichi2/anki/export/ExportDialogFragmentTest.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/test/java/com/ichi2/anki/export/ExportDialogFragmentTest.kt b/AnkiDroid/src/test/java/com/ichi2/anki/export/ExportDialogFragmentTest.kt
--- a/AnkiDroid/src/test/java/com/ichi2/anki/export/ExportDialogFragmentTest.kt	(revision ff8e1e9327f860a28449e6557d840f690c26532f)
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/export/ExportDialogFragmentTest.kt	(date 1735588795810)
@@ -13,18 +13,20 @@
  */
 package com.ichi2.anki.export
 
+import androidx.test.espresso.Espresso.onData
 import androidx.test.espresso.Espresso.onView
 import androidx.test.espresso.action.ViewActions.click
 import androidx.test.espresso.assertion.ViewAssertions.matches
 import androidx.test.espresso.matcher.RootMatchers.isDialog
+import androidx.test.espresso.matcher.RootMatchers.isPlatformPopup
 import androidx.test.espresso.matcher.ViewMatchers.Visibility
 import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
 import androidx.test.espresso.matcher.ViewMatchers.withEffectiveVisibility
 import androidx.test.espresso.matcher.ViewMatchers.withId
-import androidx.test.espresso.matcher.ViewMatchers.withText
 import androidx.test.ext.junit.runners.AndroidJUnit4
 import com.ichi2.anki.R
 import com.ichi2.testutils.AnkiFragmentScenario
+import org.hamcrest.CoreMatchers.anything
 import org.junit.Test
 import org.junit.runner.RunWith
 
@@ -32,38 +34,35 @@
 class ExportDialogFragmentTest {
     @Test
     fun colLegacyExportCheckbox() {
-        AnkiFragmentScenario.launchInContainer(ExportDialogFragment::class.java).onFragment { fragment ->
-            fragment.view?.let {
-                onView(withId(R.id.export_type_selector)).perform(click())
-                onView(withText("Anki Collection Package (.colpkg)")).perform(click())
-                onView(withId(R.id.export_legacy_checkbox_collection))
-                    .inRoot(isDialog())
-                    .check(matches(isDisplayed()))
-            }
+        AnkiFragmentScenario.launchInContainer(ExportDialogFragment::class.java).onFragment {
+            onView(withId(R.id.export_type_selector)).inRoot(isPlatformPopup()).perform(click())
+            onData(anything()).atPosition(0).perform(click())
+            onView(withId(R.id.export_legacy_checkbox_collection))
+                .inRoot(isDialog())
+                .check(matches(isDisplayed()))
         }
     }
 
     @Test
     fun apkgLegacyExportCheckbox() {
-        AnkiFragmentScenario.launchInContainer(ExportDialogFragment::class.java).onFragment { fragment ->
-            fragment.view?.let {
-                onView(withId(R.id.export_type_selector)).perform(click())
-                onView(withText("Anki Deck Package (.apkg)")).perform(click())
-                onView(withId(R.id.export_legacy_checkbox_collection))
-                    .inRoot(isDialog())
-                    .check(matches(isDisplayed()))
-            }
+        AnkiFragmentScenario.launchInContainer(ExportDialogFragment::class.java).onFragment {
+            onView(withId(R.id.export_type_selector)).inRoot(isPlatformPopup()).perform(click())
+            onData(anything()).atPosition(1).perform(click())
+            onView(withId(R.id.export_legacy_checkbox_apkg))
+                .inRoot(isDialog())
+                .check(matches(isDisplayed()))
         }
     }
 
     @Test
     fun defaultLegacyExportCheckbox() {
-        AnkiFragmentScenario.launchInContainer(ExportDialogFragment::class.java).onFragment { fragment ->
-            fragment.view?.let {
-                onView(withId(R.id.export_legacy_checkbox_collection))
-                    .inRoot(isDialog())
-                    .check(matches(withEffectiveVisibility(Visibility.GONE)))
-            }
+        AnkiFragmentScenario.launchInContainer(ExportDialogFragment::class.java).onFragment {
+            onView(withId(R.id.export_legacy_checkbox_collection))
+                .inRoot(isDialog())
+                .check(matches(withEffectiveVisibility(Visibility.GONE)))
+            onView(withId(R.id.export_legacy_checkbox_apkg)).
+                inRoot(isDialog())
+                .check(matches(withEffectiveVisibility(Visibility.GONE)))
         }
     }
 }
```

---

with this patch, scoped all to inRoot. since spinner is a platfrompopup, i used it instead of isDialog(). 

But i keep getting the errors:

<html>Matcher 'is dialog' did not match any of the following roots: [Root{application-window-token=android.view.ViewRootImpl$W@6073e236, window-token=android.view.ViewRootImpl$W@6073e236, has-window-focus=true, layout-params-type=1, layout-params-string={(0,0)(fillxfill) ty=BASE_APPLICATION wanim=0x10302fe<br/>fl=LAYOUT_IN_SCREEN LAYOUT_INSET_DECOR SPLIT_TOUCH DRAWS_SYSTEM_BAR_BACKGROUNDS<br/>pfl=NO_MOVE_ANIMATION FORCE_DRAW_STATUS_BAR_BACKGROUND FIT_INSETS_CONTROLLED<br/>bhv=DEFAULT<br/>fitSides=}, decor-view-string=DecorView{id=-1, visibility=VISIBLE, width=320, height=470, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params={(0,0)(fillxfill) ty=BASE_APPLICATION wanim=0x10302fe<br/>fl=LAYOUT_IN_SCREEN LAYOUT_INSET_DECOR SPLIT_TOUCH DRAWS_SYSTEM_BAR_BACKGROUNDS<br/>pfl=NO_MOVE_ANIMATION FORCE_DRAW_STATUS_BAR_BACKGROUND FIT_INSETS_CONTROLLED<br/>bhv=DEFAULT<br/>fitSides=}, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, child-count=1}}]

and

<html>Matcher 'with decor view of type PopupWindow$PopupDecorView' did not match any of the following roots: [Root{application-window-token=android.view.ViewRootImpl$W@11641cae, window-token=android.view.ViewRootImpl$W@11641cae, has-window-focus=true, layout-params-type=1, layout-params-string={(0,0)(fillxfill) ty=BASE_APPLICATION wanim=0x10302fe<br/>fl=LAYOUT_IN_SCREEN LAYOUT_INSET_DECOR SPLIT_TOUCH DRAWS_SYSTEM_BAR_BACKGROUNDS<br/>pfl=NO_MOVE_ANIMATION FORCE_DRAW_STATUS_BAR_BACKGROUND FIT_INSETS_CONTROLLED<br/>bhv=DEFAULT<br/>fitSides=}, decor-view-string=DecorView{id=-1, visibility=VISIBLE, width=320, height=470, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params={(0,0)(fillxfill) ty=BASE_APPLICATION wanim=0x10302fe<br/>fl=LAYOUT_IN_SCREEN LAYOUT_INSET_DECOR SPLIT_TOUCH DRAWS_SYSTEM_BAR_BACKGROUNDS<br/>pfl=NO_MOVE_ANIMATION FORCE_DRAW_STATUS_BAR_BACKGROUND FIT_INSETS_CONTROLLED<br/>bhv=DEFAULT<br/>fitSides=}, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, child-count=1}}]

---

Have tried different things, but these issues are persistenet. Is it wrong to check with withEffectiveVisibility().
and for the spinner I have no idea where I am going wrong.

---

@david-allison any idea why the roots are not matching? Are they not in scope or something?